pkg/agent/pipeline.go

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2019 Red Hat and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Network Pipeline implementation for Nimbess Agent
+
+package agent
+
+import (
+	"errors"
+	"fmt"
+	"github.com/mohae/deepcopy"
+	"reflect"
+	"sync"
+
+	"github.com/nimbess/nimbess-agent/pkg/drivers"
+	"github.com/nimbess/nimbess-agent/pkg/network"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// NimbessPipeline implements a Pipeline of network functions for packet traversal
+type NimbessPipeline struct {
+	Name        string
+	Mu          *sync.Mutex
+	Modules     []network.PipelineModule
+	Driver      drivers.Driver
+	EgressPorts []*network.EgressPort
+	MetaKey     string
+}
+
+// GetModule returns the module
+func (s *NimbessPipeline) GetModule(name string) network.PipelineModule {
+	for _, mod := range s.Modules {
+
+		if mod.GetName() == name {
+			return mod
+		}
+	}
+
+	return nil
+}
+
+// AddPort adds a port to a pipeline.
+// It returns a pointer to a EgressPort module to update other pipelines with
+func (s *NimbessPipeline) AddPort(name string, port *network.Port) (*network.EgressPort, error) {
+	if !port.Virtual {
+		return nil, errors.New("physical ports are not yet supported")
+	}
+	log.Infof("Adding new port %v to pipeline: %v", port, s.Name)
+	portMod := network.Module{Name: fmt.Sprintf("%s_ingress", name),
+		EGates: make(map[network.Gate]network.PipelineModule)}
+	// Create Ingress Port
+	iPort := &network.IngressPort{Port: port, Module: portMod}
+	s.Modules[0] = iPort
+	// Set forwarder linkage
+	// Module at index 1 should always be first meta module and already be set
+	gate := iPort.Connect(s.Modules[1], true, nil)
```

Comment:
or a port really isn't a module but this software needs the ingress port at [0]?

Reply:
right a Port isn't a module, but an IngressPort (iPort) is. In a port pipeline the first module is always an ingress port. So this code is creating an IngressPort module and then saying the first module in the pipeline should be me (index 0). Then it is connecting its gate linkage to the next module in the pipeline (index 1).